File: go/spore-core/tools/subagent_consult_test.go
package tools

import (
	"context"
	"encoding/json"
	"strings"
	"sync"
	"testing"

	sporecore "github.com/squirrelsoft-dev/spore-core/go/spore-core"
)

// ============================================================================
// Mid-loop consult mediation — seam A1 (issue #114). Orchestrator-side rules:
// R2 (mediate, not bubble), R3 (route by kind, no parent model, parent sees
// Success), R4 (per-kind budget), R5a (SoftFail → BudgetExhausted resume),
// R5b (EscalateToHuman → WaitingForHuman), R6 (no-kind / no-handlers →
// Escalate), R7 (depth-1: the handler is the orchestrator's direct child).
// ============================================================================

// recordingHandler records each instruction it is run with and returns a fixed
// answer — used to assert depth-1 routing (R3/R7).
type recordingHandler struct {
	answer string
	mu     sync.Mutex
	seen   []string
}

func (h *recordingHandler) Run(_ context.Context, opts sporecore.HarnessRunOptions) sporecore.RunResult {
	h.mu.Lock()
	h.seen = append(h.seen, opts.Task.Instruction)
	h.mu.Unlock()
	return sporecore.RunResult{Kind: sporecore.RunSuccess, Output: h.answer}
}
func (h *recordingHandler) Resume(context.Context, sporecore.PausedState, sporecore.HumanResponse, sporecore.StreamSink) sporecore.RunResult {
	return sporecore.RunResult{Kind: sporecore.RunFailure, Reason: sporecore.HaltReason{Kind: sporecore.HaltHumanHalted}}
}
func (h *recordingHandler) ResumeConsult(context.Context, sporecore.PausedState, sporecore.ConsultResponse, sporecore.StreamSink) sporecore.RunResult {
	return sporecore.RunResult{Kind: sporecore.RunFailure, Reason: sporecore.HaltReason{Kind: sporecore.HaltHumanHalted}}
}
func (h *recordingHandler) seenCount() int {
	h.mu.Lock()
	defer h.mu.Unlock()
	return len(h.seen)
}

func consultPaused() *sporecore.PausedState {
	return &sporecore.PausedState{
		SessionID:  "worker",
		TaskID:     "t",
		TurnNumber: 1,
		PendingToolCalls: []sporecore.ToolCall{
			{ID: "consult-call", Name: "ask_advice", Input: json.RawMessage(`{"kind":"advice"}`)},
		},
		Task: sporecore.NewTask("audit", "worker", sporecore.LoopStrategy{Kind: sporecore.StrategyReAct, MaxIterations: 4}),
	}
}

func consultRequest(kind string) *sporecore.ConsultRequest {
	return &sporecore.ConsultRequest{Kind: kind, Situation: "drowning", Attempts: 2, Question: "what now?"}
}

func consultResult(kind string) sporecore.RunResult {
	return sporecore.RunResult{
		Kind:           sporecore.RunConsult,
		ConsultRequest: consultRequest(kind),
		State:          consultPaused(),
		SessionID:      "worker",
		Turns:          1,
	}
}

func consultHandlers(kind, answer string, budget uint32, overflow sporecore.ConsultOverflowPolicyKind, h *recordingHandler) map[string]sporecore.ConsultHandlerEntry {
	return map[string]sporecore.ConsultHandlerEntry{
		kind: {
			Handler:  h,
			Budget:   budget,
			Overflow: sporecore.ConsultOverflowPolicy{Kind: overflow},
		},
	}
}

// R2/R3: child Consult is MEDIATED here (not bubbled). With a registered
// handler, the handler runs (no parent model), the worker is resumed, and the
// parent ultimately sees Success.
func TestConsultIsMediatedAndResumedToSuccess(t *testing.T) {
	handler := &recordingHandler{answer: "try plan B"}
	// First Run() => Consult; ResumeConsult (default scripted) => Success.
	h := &scriptedHarness{results: []sporecore.RunResult{consultResult("advice")}}
	s := newSubagent(t, h, Isolated{}).
		WithConsultHandlers(consultHandlers("advice", "try plan B", 3, sporecore.ConsultOverflowSoftFail, handler))

	r := s.Execute(context.Background(), subagentCall(map[string]any{"instruction": "x"}), sporecore.AllowAllSandbox{}, nil)
	if r.Kind != sporecore.ToolOutputSuccess || r.Content != "child done after consult" {
		t.Fatalf("expected mediated Success, got %+v", r)
	}
	// R3/R7: the handler ran exactly once, on the rendered consult request.
	if handler.seenCount() != 1 {
		t.Fatalf("handler ran %d times, want 1", handler.seenCount())
	}
	if !strings.Contains(handler.seen[0], "advice") || !strings.Contains(handler.seen[0], "what now?") {
		t.Fatalf("handler instruction missing context: %q", handler.seen[0])
	}
	// R3: worker resumed with the handler's answer.
	if len(h.resumeLog) != 1 || h.resumeLog[0].Kind != sporecore.ConsultRespAnswer || h.resumeLog[0].Text != "try plan B" {
		t.Fatalf("expected resume with Answer 'try plan B', got %+v", h.resumeLog)
	}
}

// R4 + R5a: handler runs up to budget times; the (budget+1)th consult overflows.
// With SoftFail, the worker is resumed with BudgetExhausted and finishes.
func TestBudgetOverflowSoftFailResumesWithBudgetExhausted(t *testing.T) {
	handler := &recordingHandler{answer: "advice answer"}
	// budget = 1: Run() => Consult; ResumeConsult => Consult again (over budget);
	// ResumeConsult => Success.
	h := &scriptedHarness{results: []sporecore.RunResult{
		consultResult("advice"),
		consultResult("advice"),
	}}
	s := newSubagent(t, h, Isolated{}).
		WithConsultHandlers(consultHandlers("advice", "advice answer", 1, sporecore.ConsultOverflowSoftFail, handler))

	r := s.Execute(context.Background(), subagentCall(map[string]any{"instruction": "x"}), sporecore.AllowAllSandbox{}, nil)
	if r.Kind != sporecore.ToolOutputSuccess {
		t.Fatalf("expected Success, got %+v", r)
	}
	// R4: handler ran exactly once (budget = 1).
	if handler.seenCount() != 1 {
		t.Fatalf("handler ran %d times, want 1", handler.seenCount())
	}
	// R5a: first resume = Answer, second resume = BudgetExhausted.
	if len(h.resumeLog) != 2 {
		t.Fatalf("expected 2 resumes, got %d", len(h.resumeLog))
	}
	if h.resumeLog[0].Kind != sporecore.ConsultRespAnswer {
		t.Fatalf("first resume should be Answer, got %q", h.resumeLog[0].Kind)
	}
	if h.resumeLog[1].Kind != sporecore.ConsultRespBudgetExhausted {
		t.Fatalf("second resume should be BudgetExhausted, got %q", h.resumeLog[1].Kind)
	}
}

// R5b: budget overflow with EscalateToHuman → ToolOutput.WaitingForHuman.
func TestBudgetOverflowEscalateToHuman(t *testing.T) {
	handler := &recordingHandler{answer: "x"}
	// budget = 0: the FIRST consult is already over budget → escalate.
	h := &scriptedHarness{results: []sporecore.RunResult{consultResult("advice")}}
	s := newSubagent(t, h, Isolated{}).
		WithConsultHandlers(consultHandlers("advice", "x", 0, sporecore.ConsultOverflowEscalateToHuman, handler))

	r := s.Execute(context.Background(), subagentCall(map[string]any{"instruction": "x"}), sporecore.AllowAllSandbox{}, nil)
	if r.Kind != sporecore.ToolOutputWaitingForHuman {
		t.Fatalf("expected WaitingForHuman, got %+v", r)
	}
	if r.ChildState == nil || r.ChildState.ParentToolCallID != "parent-call-1" {
		t.Fatalf("bad child state: %+v", r.ChildState)
	}
	if r.Request == nil || r.Request.Kind != sporecore.HumanReqReview {
		t.Fatalf("expected Review human request, got %+v", r.Request)
	}
	// Handler never ran (over budget from the start).
	if handler.seenCount() != 0 {
		t.Fatalf("handler should not have run, ran %d", handler.seenCount())
	}
}

// R6: a consult with NO matching handler (map present, wrong kind) →
// ToolOutput.Escalate (loud, not silent).
func TestConsultNoMatchingKindEscalates(t *testing.T) {
	handler := &recordingHandler{answer: "x"}
	h := &scriptedHarness{results: []sporecore.RunResult{consultResult("research")}}
	s := newSubagent(t, h, Isolated{}).
		WithConsultHandlers(consultHandlers("advice", "x", 3, sporecore.ConsultOverflowSoftFail, handler))

	r := s.Execute(context.Background(), subagentCall(map[string]any{"instruction": "x"}), sporecore.AllowAllSandbox{}, nil)
	if r.Kind != sporecore.ToolOutputEscalate {
		t.Fatalf("expected Escalate, got %+v", r)
	}
	if r.Signal == nil || r.Signal.Kind != sporecore.SignalAbort || !strings.Contains(r.Signal.Reason, "research") {
		t.Fatalf("expected Abort mentioning research, got %+v", r.Signal)
	}
}

// R6 (degradation, R9): with NO handlers installed at all, a child consult is
// treated as the no-matching-kind case → Escalate. Existing callers that never
// set ConsultHandlers are therefore unaffected — their subagents simply do not
// mediate.
func TestConsultWithNoHandlersEscalates(t *testing.T) {
	h := &scriptedHarness{results: []sporecore.RunResult{consultResult("advice")}}
	s := newSubagent(t, h, Isolated{})
	r := s.Execute(context.Background(), subagentCall(map[string]any{"instruction": "x"}), sporecore.AllowAllSandbox{}, nil)
	if r.Kind != sporecore.ToolOutputEscalate {
		t.Fatalf("expected Escalate, got %+v", r)
	}
}

// R7 (depth-1): the handler harness is run as the ORCHESTRATOR's direct child —
// it is never installed under the worker's registry, and SubagentTool itself
// still reports IsSubagentTool so the depth-1 construction guard remains intact.
func TestConsultHandlerRunsAsOrchestratorChild(t *testing.T) {
	handler := &recordingHandler{answer: "ans"}
	h := &scriptedHarness{results: []sporecore.RunResult{consultResult("advice")}}
	s := newSubagent(t, h, Isolated{}).
		WithConsultHandlers(consultHandlers("advice", "ans", 3, sporecore.ConsultOverflowSoftFail, handler))
	_ = s.Execute(context.Background(), subagentCall(map[string]any{"instruction": "x"}), sporecore.AllowAllSandbox{}, nil)
	// The handler ran directly (its Run was invoked), proving it is the
	// orchestrator's direct child rather than nested under the worker.
	if handler.seenCount() != 1 {
		t.Fatalf("handler should run once as orchestrator child, ran %d", handler.seenCount())
	}
	if !s.IsSubagentTool() {
		t.Fatal("SubagentTool must still report IsSubagentTool for the depth-1 guard")
	}
}

File: typescript/packages/tools/tests/subagent-consult.test.ts
/**
 * SubagentTool consult-mediation tests (spore-core issue #114) — mirror
 * `rust/crates/spore-core/src/tools/subagent.rs#tests` (the consult block).
 *
 * The SubagentTool drives the FULL consult cycle internally (seam A1): on a
 * child `RunResult.consult` it routes by `kind`, enforces the per-kind budget,
 * runs the handler as the orchestrator's direct child, and resumes the worker —
 * the parent orchestrator's model never sees the consult.
 *
 * Rules covered: R2 (mediate, not bubble), R3 (route by kind, no parent model,
 * parent sees success), R4 (per-kind budget), R5a (soft_fail → budget_exhausted),
 * R5b (escalate_to_human → waiting_for_human), R6 (no matching kind → escalate;
 * no handlers → escalate), R7 (depth-1: the handler runs via handler.run).
 */

import {
  emptyAggregateUsage,
  emptyBudgetSnapshot,
  emptySessionState,
  harnessTesting,
  newTask,
  SessionId,
  TaskId,
  toolRegistry,
  type ConsultHandlerEntry,
  type ConsultOverflowPolicy,
  type ConsultRequest,
  type ConsultResponse,
  type Harness,
  type HarnessRunOptions,
  type HumanResponse,
  type PausedState,
  type RunResult,
  type StreamSink,
  type ToolCall,
} from "@spore/core";
import { describe, expect, it } from "vitest";

import { SubagentTool } from "../src/index.js";

const { AllowAllSandbox } = harnessTesting;
const { StandardToolRegistry, toolRegistryMock } = toolRegistry;
const ctx = toolRegistryMock.testCtx();

/** Scripted harness whose `resumeConsult` logs the response and pops the next
 *  scripted RunResult — so a test can assert exactly what the worker was
 *  resumed with. An empty queue resolves to a terminal success. */
class ScriptedHarness implements Harness {
  readonly resumeLog: ConsultResponse[] = [];
  constructor(private readonly results: RunResult[]) {}

  async run(_opts: HarnessRunOptions): Promise<RunResult> {
    return this.pop();
  }
  async resume(
    _state: PausedState,
    _response: HumanResponse,
    _onStream?: StreamSink,
  ): Promise<RunResult> {
    return this.terminal();
  }
  async resumeConsult(
    _state: PausedState,
    response: ConsultResponse,
    _onStream?: StreamSink,
  ): Promise<RunResult> {
    this.resumeLog.push(response);
    return this.pop();
  }
  private pop(): RunResult {
    const r = this.results.shift();
    return r ?? this.terminal();
  }
  private terminal(): RunResult {
    return {
      kind: "success",
      output: "child done after consult",
      session_id: SessionId.of("worker"),
      usage: emptyAggregateUsage(),
      turns: 1,
    };
  }
}

/** Handler harness that records each instruction it is run with and returns a
 *  fixed answer. Used to assert depth-1 routing (R3/R7). */
class RecordingHandler implements Harness {
  readonly seen: string[] = [];
  constructor(private readonly answer: string) {}
  async run(opts: HarnessRunOptions): Promise<RunResult> {
    this.seen.push(opts.task.instruction);
    return {
      kind: "success",
      output: this.answer,
      session_id: SessionId.of("handler"),
      usage: emptyAggregateUsage(),
      turns: 1,
    };
  }
  async resume(
    _state: PausedState,
    _response: HumanResponse,
    _onStream?: StreamSink,
  ): Promise<RunResult> {
    return {
      kind: "failure",
      reason: { kind: "human_halted" },
      session_id: SessionId.of("handler"),
      usage: emptyAggregateUsage(),
      turns: 0,
    };
  }
}

function consultPaused(): PausedState {
  const sessionId = SessionId.of("worker");
  return {
    session_id: sessionId,
    task_id: TaskId.of("t"),
    turn_number: 1,
    session_state: emptySessionState(),
    pending_tool_calls: [
      { id: "consult-call", name: "ask_advice", input: { kind: "advice" } },
    ],
    approved_results: [],
    task: newTask("audit", sessionId, { kind: "re_act", max_iterations: 4 }),
    budget_used: emptyBudgetSnapshot(),
    child_state: null,
  };
}

function consultRequest(kind: string): ConsultRequest {
  return { kind, situation: "drowning", attempts: 2, question: "what now?" };
}

function consultResult(kind: string): RunResult {
  return {
    kind: "consult",
    request: consultRequest(kind),
    state: consultPaused(),
    session_id: SessionId.of("worker"),
    usage: emptyAggregateUsage(),
    turns: 1,
  };
}

function handlers(
  kind: string,
  handler: RecordingHandler,
  budget: number,
  overflow: ConsultOverflowPolicy,
): Map<string, ConsultHandlerEntry> {
  return new Map([[kind, { handler, budget, overflow }]]);
}

function callWith(input: unknown, id = "parent-call-1"): ToolCall {
  return { id, name: "subagent", input };
}

function buildSubagent(
  harness: Harness,
  consultHandlers?: Map<string, ConsultHandlerEntry>,
): SubagentTool {
  return SubagentTool.buildOrThrow({
    name: "subagent",
    description: "child",
    inputSchema: { type: "object" },
    timeoutMs: 5_000,
    contextSharing: { kind: "isolated" },
    harness,
    childRegistry: new StandardToolRegistry(),
    consultHandlers,
  });
}

describe("SubagentTool consult mediation (#114)", () => {
  // R2/R3: child consult is MEDIATED here (not bubbled). With a registered
  // handler, the handler runs (no parent model), the worker is resumed, and the
  // parent ultimately sees success.
  it("R2/R3: mediates and resumes to success without bubbling", async () => {
    const handler = new RecordingHandler("try plan B");
    const harness = new ScriptedHarness([consultResult("advice")]);
    const sub = buildSubagent(
      harness,
      handlers("advice", handler, 3, { kind: "soft_fail" }),
    );
    const r = await sub.execute(
      callWith({ instruction: "x" }),
      new AllowAllSandbox(),
      ctx,
    );

    // R3: parent sees success (the consult never reached its model).
    expect(r.kind).toBe("success");
    if (r.kind === "success")
      expect(r.content).toBe("child done after consult");
    // R3/R7: the handler ran exactly once, on the rendered consult request.
    expect(handler.seen.length).toBe(1);
    expect(handler.seen[0]).toContain("advice");
    expect(handler.seen[0]).toContain("what now?");
    // R3: worker resumed with the handler's answer.
    expect(harness.resumeLog.length).toBe(1);
    expect(harness.resumeLog[0]).toEqual({
      kind: "answer",
      text: "try plan B",
    });
  });

  // R4 + R5a: handler runs up to `budget` times; the (budget+1)th consult
  // overflows. With soft_fail, the worker is resumed with budget_exhausted.
  it("R4/R5a: budget overflow soft_fail resumes with budget_exhausted", async () => {
    const handler = new RecordingHandler("advice answer");
    // budget = 1: run → consult; resumeConsult → consult again (over budget);
    // resumeConsult → success (queue empty).
    const harness = new ScriptedHarness([
      consultResult("advice"),
      consultResult("advice"),
    ]);
    const sub = buildSubagent(
      harness,
      handlers("advice", handler, 1, { kind: "soft_fail" }),
    );
    const r = await sub.execute(
      callWith({ instruction: "x" }),
      new AllowAllSandbox(),
      ctx,
    );

    expect(r.kind).toBe("success");
    // R4: handler ran exactly once (budget = 1).
    expect(handler.seen.length).toBe(1);
    // R5a: first resume = answer, second resume = budget_exhausted.
    expect(harness.resumeLog.length).toBe(2);
    expect(harness.resumeLog[0]!.kind).toBe("answer");
    expect(harness.resumeLog[1]!.kind).toBe("budget_exhausted");
  });

  // R5b: budget overflow with escalate_to_human → ToolOutput.waiting_for_human.
  it("R5b: budget overflow escalate_to_human surfaces waiting_for_human", async () => {
    const handler = new RecordingHandler("x");
    // budget = 0: the FIRST consult is already over budget → escalate.
    const harness = new ScriptedHarness([consultResult("advice")]);
    const sub = buildSubagent(
      harness,
      handlers("advice", handler, 0, { kind: "escalate_to_human" }),
    );
    const r = await sub.execute(
      callWith({ instruction: "x" }),
      new AllowAllSandbox(),
      ctx,
    );

    expect(r.kind).toBe("waiting_for_human");
    if (r.kind === "waiting_for_human") {
      expect(r.child_state.parent_tool_call_id).toBe("parent-call-1");
      expect(r.request.kind).toBe("review");
    }
    // Handler never ran (over budget from the start).
    expect(handler.seen.length).toBe(0);
  });

  // R6: a consult with NO matching handler (map present, wrong kind) → escalate.
  it("R6: consult with no matching kind escalates", async () => {
    const handler = new RecordingHandler("x");
    const harness = new ScriptedHarness([consultResult("research")]);
    const sub = buildSubagent(
      harness,
      handlers("advice", handler, 3, { kind: "soft_fail" }),
    );
    const r = await sub.execute(
      callWith({ instruction: "x" }),
      new AllowAllSandbox(),
      ctx,
    );

    expect(r.kind).toBe("escalate");
    if (r.kind === "escalate") {
      expect(r.signal.kind).toBe("abort");
      if (r.signal.kind === "abort")
        expect(r.signal.reason).toContain("research");
    }
  });

  // R6 (degradation): with NO handlers installed at all, a child consult is
  // treated as the no-matching-kind case → escalate.
  it("R6: consult with no handlers installed escalates", async () => {
    const harness = new ScriptedHarness([consultResult("advice")]);
    const sub = buildSubagent(harness);
    const r = await sub.execute(
      callWith({ instruction: "x" }),
      new AllowAllSandbox(),
      ctx,
    );
    expect(r.kind).toBe("escalate");
  });
});
